### ISSUE TITLE
Training crashes on second epoch with "Unable to unpickle data from other workers"

### ISSUE THREAD

**Issue by davidecaroselli** [pytorch/fairseq#780]
Hello team,

I would like to ask more info on an error that I hit when training on my 8GPUs machine. The problem is that the training crashes when it is almost at half of the second epoch, this is the interesting part of training log:
```
| epoch 002:  15200 / 41687 loss=3.211, nll_loss=1.497, ppl=2.82, wps=33227, ups=0, wpb=81220.032, bsz=4526.620, num_updates=56888, lr=0.000132584, gnorm=0.213, clip=0.000, oom=0.000, wall=141888, train_wall=135650
| epoch 002:  15300 / 41687 loss=3.211, nll_loss=1.497, ppl=2.82, wps=33230, ups=0, wpb=81221.187, bsz=4526.611, num_updates=56988, lr=0.000132467, gnorm=0.213, clip=0.000, oom=0.000, wall=142130, train_wall=135888
Traceback (most recent call last):
  File "/home/ubuntu/.local/bin/fairseq-train", line 11, in <module>
    load_entry_point('fairseq', 'console_scripts', 'fairseq-train')()
  File "/home/ubuntu/workspace/fairseq/fairseq_cli/train.py", line 403, in cli_main
    nprocs=args.distributed_world_size,
  File "/home/ubuntu/.local/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 167, in spawn
    while not spawn_context.join():
  File "/home/ubuntu/.local/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 114, in join
    raise Exception(msg)
Exception: 

-- Process 5 terminated with the following error:
Traceback (most recent call last):
  File "/home/ubuntu/workspace/fairseq/fairseq/distributed_utils.py", line 151, in all_gather_list
    result.append(pickle.loads(bytes(out_buffer[2 : size + 2].tolist())))
_pickle.UnpicklingError: invalid load key, '\x16'.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/.local/lib/python3.6/site-packages/torch/multiprocessing/spawn.py", line 19, in _wrap
    fn(i, *args)
  File "/home/ubuntu/workspace/fairseq/fairseq_cli/train.py", line 380, in distributed_main
    main(args, init_distributed=True)
  File "/home/ubuntu/workspace/fairseq/fairseq_cli/train.py", line 106, in main
    train(args, trainer, task, epoch_itr)
  File "/home/ubuntu/workspace/fairseq/fairseq_cli/train.py", line 141, in train
    log_output = trainer.train_step(samples)
  File "/home/ubuntu/workspace/fairseq/fairseq/trainer.py", line 221, in train_step
    [logging_outputs, sample_sizes, ooms, self._prev_grad_norm],
  File "/home/ubuntu/workspace/fairseq/fairseq/distributed_utils.py", line 155, in all_gather_list
    'Unable to unpickle data from other workers. all_gather_list requires all '
Exception: Unable to unpickle data from other workers. all_gather_list requires all workers to enter the function together, so this error usually indicates that the workers have fallen out of sync somehow. Workers can fall out of sync if one of them runs out of memory, or if there are other conditions in your training script that can cause one worker to finish an epoch while other workers are still iterating over their portions of the data.
```

The error is **totally deterministic**: if I re-run the training, it start from checkpoint 56000 and it will crash again after update `56988`.

So my understanding is that one of the workers run out of GPU RAM and crashes. If this is the problem, how can it be possible? If the training has already seen the whole dataset (because no error has been thrown during first epoch), how can it crash on the same dataset?

Given this evidence, we suppose that, while the dataset is exactly the same, the batches are created differently for every epoch (but in a deterministic way!). So here, we are unlucky, and a new dataset created at epoch 2 is a combination of sentence pairs that is too large for our GPU RAM. Is this possible or are we're barking up the wrong tree?

Thanks for your help!

**Comment by myleott** [comment]
The order of the batches are completely deterministic for a given seed/epoch, so it’s good news that resuming from a checkpoint fails at the same time :) This is likely an OOM, which can sometimes happen if there is memory fragmentation due to PuTorch’s caching memory allocator. Can you try resuming training using a different --seed? If it works, then it’s probably just an unlucky arrangement of batches that causes OOM due to fragmentation. If it fails, then maybe something else is wrong...

**Comment by davidecaroselli** [comment]
Hi @myleott turns out you're right!
I copied the data on a V100 GPU machine and restarted the training just for the time of 1 more checkpoint. I noticed that for 2 GPUs the RAM was above 12G, that explains why training crashed on my 11G setup.

After the new checkpoint has been saved, I moved all back to my original setup and restarted the training. Everything works as expected!

Thanks for you help, I now understand a bit more how fairseq works internally :)
